src/main/scala/mesosphere/mesos/DiskResourceMatcher.scala

```diff
@@ -0,0 +1,203 @@
+package mesosphere.mesos
+
+import mesosphere.marathon.state.{ PersistentVolume, DiskType, DiskSource }
+import mesosphere.mesos.protos.Resource
+import org.apache.mesos.Protos
+import org.apache.mesos.Protos.Resource.DiskInfo.Source
+import mesosphere.mesos.ResourceMatcher.{ ResourceSelector, consumeResources }
+import scala.annotation.tailrec
+import scala.collection.immutable
+
+/**
+  * Match volumes against disk resources and return results which keep disk sources and persistentVolumes associated.
+  *
+  * TODO - handle matches for a single volume across multiple resource offers for the same disk
+  */
+class DiskResourceMatcher(
+    selector: ResourceSelector,
+    scratchDisk: Double,
+    volumes: Iterable[PersistentVolume],
+    scope: ScalarMatchResult.Scope = ScalarMatchResult.Scope.NoneDisk) extends ResourceMatcher {
+
+  val resourceName = Resource.DISK
+
+  import ResourceUtil.RichResource
+  private[mesos] case class SourceResources(source: Option[Source], resources: List[Protos.Resource]) {
+    lazy val size = resources.foldLeft(0.0)(_ + _.getScalar.getValue)
+  }
+  private[mesos] object SourceResources extends ((Option[Source], List[Protos.Resource]) => SourceResources) {
+    def listFromResources(l: List[Protos.Resource]): List[SourceResources] = {
+      l.groupBy(_.diskSourceOption).map(SourceResources.tupled).toList
+    }
+  }
+
+  /*
+   Prioritize resources to make the most sensible allocation.
+   - If requesting full disk, allocate the smallest disk volume that meets constraints
+   - If requesting root, just return it because there can only be one.
+   - If requesting path disk, allocate the largest volume possible to spread allocation evenly.
+   This may not be ideal if you'd prefer to leave room for larger allocations instead.
+
+   TODO - test. Also, parameterize?
+   */
+  private def prioritizeDiskResources(
+    diskType: DiskType,
+    resources: List[SourceResources]): List[SourceResources] = {
+    diskType match {
+      case DiskType.Root =>
+        resources
+      case DiskType.Path =>
+        resources.sortBy(_.size)(implicitly[Ordering[Double]].reverse)
```

Comment:
or .. nevermind since this is just shuffled code; we could address in a follow-up PR